File: Python/112. Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def hasPathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: bool
        """
        def dfs(node, s, val):
            if node is None:
                return False
            s += node.val
            if node.left is None and node.right is None:
                return s == val
            return dfs(node.left, s, val) or dfs(node.right, s, val)
        return dfs(root, 0, sum)

File: C++/112. Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool dfs(TreeNode* node, int sum, int target)
    {
        sum += node->val;
        if(sum == target && !node->left && !node->right)
            return true;
        if(node->left && dfs(node->left, sum, target))
            return true;
        if(node->right && dfs(node->right, sum, target))
            return true;
        return false;
            
        
    }
    bool hasPathSum(TreeNode* root, int sum) {
        if(root == nullptr)
            return false;
        return dfs(root, 0, sum);
    }
};